### PR TITLE
feat: postfix `rolling` expression as a special case of window functions.

### PR DESCRIPTION
**Pull request by ritchie46**
```python
  >>> dates = [
  ...     "2020-01-01 13:45:48",
  ...     "2020-01-01 16:42:13",
  ...     "2020-01-01 16:45:09",
  ...     "2020-01-02 18:12:48",
  ...     "2020-01-03 19:45:32",
  ...     "2020-01-08 23:16:43",
  ... ]
  >>> df = pl.DataFrame({"dt": dates, "a": [3, 7, 5, 9, 2, 1]}).with_columns(
  ...     pl.col("dt").str.strptime(pl.Datetime).set_sorted()
  ... )
  >>> df.with_columns(
  ...     sum_a = pl.sum("a").rolling(index_column="dt", period="2d"),
  ...     min_a = pl.min("a").rolling(index_column="dt", period="2d"),
  ...     max_a = pl.max("a").rolling(index_column="dt", period="2d"),
  ... )
  shape: (6, 5)
  ┌─────────────────────┬─────┬───────┬───────┬───────┐
  │ dt                  ┆ a   ┆ sum_a ┆ min_a ┆ max_a │
  │ ---                 ┆ --- ┆ ---   ┆ ---   ┆ ---   │
  │ datetime[μs]        ┆ i64 ┆ i64   ┆ i64   ┆ i64   │
  ╞═════════════════════╪═════╪═══════╪═══════╪═══════╡
  │ 2020-01-01 13:45:48 ┆ 3   ┆ 3     ┆ 3     ┆ 3     │
  │ 2020-01-01 16:42:13 ┆ 7   ┆ 10    ┆ 3     ┆ 7     │
  │ 2020-01-01 16:45:09 ┆ 5   ┆ 15    ┆ 3     ┆ 7     │
  │ 2020-01-02 18:12:48 ┆ 9   ┆ 24    ┆ 3     ┆ 9     │
  │ 2020-01-03 19:45:32 ┆ 2   ┆ 11    ┆ 2     ┆ 9     │
  │ 2020-01-08 23:16:43 ┆ 1   ┆ 1     ┆ 1     ┆ 1     │
  └─────────────────────┴─────┴───────┴───────┴───────┘
```

For cases of `group_by_rolling` that don't require a traditional `group by` (which reorders other columns) we can now use a `rolling` expression to get the rolling window functionality.

Additionally I want to introduce a new special syntax for `over` and `rolling` and that is a grouped expression that will be expanded. E.g.

The input:
```python
(
   pl.sum("a").alias("sum_a"),
   pl.min("a").alias("min_a"),
   pl.max("a").alias("max_a"),
).rolling(index_column="t")
```

Should expand to:

```python
[
   pl.sum("a").alias("sum_a").rolling(index_column="t"),
   pl.min("a").alias("min_a").rolling(index_column="t"),
   pl.max("a").alias("max_a").rolling(index_column="t"),
]
```

And similarly for the `over` case.

### CODE REVIEW

File: py-polars/polars/expr/expr.py [hunk 1]
```diff
@@ -3180,6 +3180,125 @@ def over(
...
+            {Int32, Int64}. Note that Int32 gets temporarily cast to Int64, so if
+            performance matters use an Int64 column.
+        period
+            length of the window - must be non-negative
+        offset
+            offset of the window. Default is -period
+        closed : {'right', 'left', 'both', 'none'}
+            Define which sides of the temporal interval are closed (inclusive).
+        check_sorted
+            When the ``by`` argument is given, polars can not check sortedness
```
Comment (MarcoGorelli) [inline py-polars/polars/expr/expr.py:3253]: there's no `by` argument here?

**Review by MarcoGorelli — COMMENTED** [review]
Wow, truly wonderful! I guess this would make `Expr.rolling_mean` (and friends) irrelevant, and they can be deprecated?

**Comment by lucazanna** [comment]
Look forward to trying this 

As the concept is somehow similar, will there be a cumulative method also ?

**Comment by ritchie46** [comment]
> Wow, truly wonderful! I guess this would make `Expr.rolling_mean` (and friends) irrelevant, and they can be deprecated?

Not exactly as those have different windows. They are fixes size on number of slots. Where these are varying sizes depending on window duration.

**Comment by ion-elgreco** [comment]
> > Wow, truly wonderful! I guess this would make `Expr.rolling_mean` (and friends) irrelevant, and they can be deprecated?
> 
> Not exactly as those have different windows. They are fixes size on number of slots. Where these are varying sizes depending on window duration.

When would one then use `pl.col('foo').mean.rolling()` over `pl.col('foo').rolling_mean()`?

File: py-polars/polars/expr/expr.py (hunk 1 above)
Comment (ritchie46) [reply py-polars/polars/expr/expr.py:3253]: No.. because the by argument would need to reorder the other columns or this output.

**Comment by MarcoGorelli** [comment]
I might be missing something, but for the example you provided, they look equivalent?

By temporal dtype:
```python
In [22]:         >>> df.with_columns(
    ...:         ...     sum_a=pl.sum("a").rolling(index_column="dt", period="2d"),
    ...:         ...     min_a=pl.min("a").rolling(index_column="dt", period="2d"),
    ...:         ...     max_a=pl.max("a").rolling(index_column="dt", period="2d"),
    ...:         ... )
Out[22]:
shape: (6, 5)
┌─────────────────────┬─────┬───────┬───────┬───────┐
│ dt                  ┆ a   ┆ sum_a ┆ min_a ┆ max_a │
│ ---                 ┆ --- ┆ ---   ┆ ---   ┆ ---   │
│ datetime[μs]        ┆ i64 ┆ i64   ┆ i64   ┆ i64   │
╞═════════════════════╪═════╪═══════╪═══════╪═══════╡
│ 2020-01-01 13:45:48 ┆ 3   ┆ 3     ┆ 3     ┆ 3     │
│ 2020-01-01 16:42:13 ┆ 7   ┆ 10    ┆ 3     ┆ 7     │
│ 2020-01-01 16:45:09 ┆ 5   ┆ 15    ┆ 3     ┆ 7     │
│ 2020-01-02 18:12:48 ┆ 9   ┆ 24    ┆ 3     ┆ 9     │
│ 2020-01-03 19:45:32 ┆ 2   ┆ 11    ┆ 2     ┆ 9     │
│ 2020-01-08 23:16:43 ┆ 1   ┆ 1     ┆ 1     ┆ 1     │
└─────────────────────┴─────┴───────┴───────┴───────┘

In [23]:         >>> df.with_columns(
    ...:         ...     sum_a=pl.col("a").rolling_sum("2d", by="dt", closed='right'),
    ...:         ...     min_a=pl.col("a").rolling_min("2d", by="dt", closed='right'),
    ...:         ...     max_a=pl.col('a').rolling_max('2d', by='dt', closed='right'),
    ...:         ... )
Out[23]:
shape: (6, 5)
┌─────────────────────┬─────┬───────┬───────┬───────┐
│ dt                  ┆ a   ┆ sum_a ┆ min_a ┆ max_a │
│ ---                 ┆ --- ┆ ---   ┆ ---   ┆ ---   │
│ datetime[μs]        ┆ i64 ┆ i64   ┆ i64   ┆ i64   │
╞═════════════════════╪═════╪═══════╪═══════╪═══════╡
│ 2020-01-01 13:45:48 ┆ 3   ┆ 3     ┆ 3     ┆ 3     │
│ 2020-01-01 16:42:13 ┆ 7   ┆ 10    ┆ 3     ┆ 7     │
│ 2020-01-01 16:45:09 ┆ 5   ┆ 15    ┆ 3     ┆ 7     │
│ 2020-01-02 18:12:48 ┆ 9   ┆ 24    ┆ 3     ┆ 9     │
│ 2020-01-03 19:45:32 ┆ 2   ┆ 11    ┆ 2     ┆ 9     │
│ 2020-01-08 23:16:43 ┆ 1   ┆ 1     ┆ 1     ┆ 1     │
└─────────────────────┴─────┴───────┴───────┴───────┘
```

By number of slots:
```python
In [31]:         >>> df.with_columns(
    ...:         ...     sum_a=pl.sum("a").rolling(index_column="row_nr", period="2i"),
    ...:         ...     min_a=pl.min("a").rolling(index_column="row_nr", period="2i"),
    ...:         ...     max_a=pl.max("a").rolling(index_column="row_nr", period="2i"),
    ...:         ... )
Out[31]:
shape: (6, 6)
┌────────┬─────────────────────┬─────┬───────┬───────┬───────┐
│ row_nr ┆ dt                  ┆ a   ┆ sum_a ┆ min_a ┆ max_a │
│ ---    ┆ ---                 ┆ --- ┆ ---   ┆ ---   ┆ ---   │
│ i64    ┆ datetime[μs]        ┆ i64 ┆ i64   ┆ i64   ┆ i64   │
╞════════╪═════════════════════╪═════╪═══════╪═══════╪═══════╡
│ 0      ┆ 2020-01-01 13:45:48 ┆ 3   ┆ 3     ┆ 3     ┆ 3     │
│ 1      ┆ 2020-01-01 16:42:13 ┆ 7   ┆ 10    ┆ 3     ┆ 7     │
│ 2      ┆ 2020-01-01 16:45:09 ┆ 5   ┆ 12    ┆ 5     ┆ 7     │
│ 3      ┆ 2020-01-02 18:12:48 ┆ 9   ┆ 14    ┆ 5     ┆ 9     │
│ 4      ┆ 2020-01-03 19:45:32 ┆ 2   ┆ 11    ┆ 2     ┆ 9     │
│ 5      ┆ 2020-01-08 23:16:43 ┆ 1   ┆ 3     ┆ 1     ┆ 2     │
└────────┴─────────────────────┴─────┴───────┴───────┴───────┘

In [32]:         >>> df.with_columns(
    ...:         ...     sum_a=pl.col("a").rolling_sum("2i", closed='right'),
    ...:         ...     min_a=pl.col("a").rolling_min("2i", closed='right'),
    ...:         ...     max_a=pl.col('a').rolling_max('2i', closed='right'),
    ...:         ... )
Out[32]:
shape: (6, 6)
┌────────┬─────────────────────┬─────┬───────┬───────┬───────┐
│ row_nr ┆ dt                  ┆ a   ┆ sum_a ┆ min_a ┆ max_a │
│ ---    ┆ ---                 ┆ --- ┆ ---   ┆ ---   ┆ ---   │
│ i64    ┆ datetime[μs]        ┆ i64 ┆ i64   ┆ i64   ┆ i64   │
╞════════╪═════════════════════╪═════╪═══════╪═══════╪═══════╡
│ 0      ┆ 2020-01-01 13:45:48 ┆ 3   ┆ 3     ┆ 3     ┆ 3     │
│ 1      ┆ 2020-01-01 16:42:13 ┆ 7   ┆ 10    ┆ 3     ┆ 7     │
│ 2      ┆ 2020-01-01 16:45:09 ┆ 5   ┆ 12    ┆ 5     ┆ 7     │
│ 3      ┆ 2020-01-02 18:12:48 ┆ 9   ┆ 14    ┆ 5     ┆ 9     │
│ 4      ┆ 2020-01-03 19:45:32 ┆ 2   ┆ 11    ┆ 2     ┆ 9     │
│ 5      ┆ 2020-01-08 23:16:43 ┆ 1   ┆ 3     ┆ 1     ┆ 2     │
└────────┴─────────────────────┴─────┴───────┴───────┴───────┘
```

**Comment by ritchie46** [comment]
@MarcoGorelli You are right. I was under the assumption they were different. Let me check if I refactor some of the backend and then we could indeed deprecate those in favor of the `rolling` postfix.

File: py-polars/polars/expr/expr.py (hunk 1 above)
Comment (MarcoGorelli) [reply py-polars/polars/expr/expr.py:3253]: yeah sorry I meant that the docstring still says "When the ``by`` argument is given", which probably needs removing

**Comment by avimallu** [comment]
This is very powerful. Having a similar syntax for cumulative will be awesome, with fast paths for mean, sum and other familiar functions.

**Comment by ritchie46** [comment]
@MarcoGorelli the specialize `rolling_mean` etc. have arguments that we cannot generalize to the other expressions this way. E.g. `weights`, `center` and `min_elements` are specialized for that implementation. I believe these arguments are too useful to deprecate the specialized variants. :thinking:

**Comment by MarcoGorelli** [comment]
ah yeah, true that

**Comment by ritchie46** [comment]
> ah yeah, true that

I will still try to make sure we can hit the specialized execution paths with the new cool syntax though.

**Comment by thomasfrederikhoeck** [comment]
> @MarcoGorelli the specialize `rolling_mean` etc. have arguments that we cannot generalize to the other expressions this way. E.g. `weights`, `center` and `min_elements` are specialized for that implementation. I believe these arguments are too useful to deprecate the specialized variants. :thinking: 

Wouldn't it be useful for min_elements to be part of .rolling() in general?

**Comment by ritchie46** [comment]
> Wouldn't it be useful for min_elements to be part of .rolling() in general?

Then all our aggregations should be made aware of this parameter, which is a lot of work and I am not certain we should have that in the first place as we are then combining compute goals. The fact that we cannot merge these functionalities proofs that having multiple compute options in a single operator complicates things.

> When would one then use pl.col('foo').mean().rolling() over pl.col('foo').rolling_mean()?

At the moment `rolling_mean` is specialized with some extra arguments and it can be rolling without the requirement of sorted index column.

**Pull request merged**